File: scala/queen-attack/src/main/scala/QueenAttack.scala
object QueenAttack {
  def canAttack(queen: Queen, other: Queen) = {
    val dx = (queen.x - other.x).abs
    val dy = (queen.y - other.y).abs

    dx == 0 || dy == 0 || dx == dy

  }

}

object Queen {

  def create(x: Int, y: Int): Option[Queen] = {
    if (x >= 0 && x < 8 && y >= 0 && y < 8) {
      Some(Queen(x, y))
    } else {
      None
    }
  }

}

case class Queen(x: Int, y: Int)

File: erlang/queen-attack/src/queen_attack.erl
-module(queen_attack).

-export([can_attack/2]).

can_attack({X, _}, {X, _}) -> true;
can_attack({_, Y}, {_, Y}) -> true;
can_attack({X1, Y1}, {X2, Y2}) -> abs(X1 - X2) == abs(Y1 - Y2).

